File: detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/stream/Streams.java
package io.buildfoundation.bazel.detekt.stream;

import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintStream;

public interface Streams {

    InputStream input();
    PrintStream output();
    PrintStream error();

    final class Impl implements Streams {

        private final InputStream input;
        private final PrintStream output;
        private final PrintStream error;

        public Impl() {
            this.input = System.in;
            this.output = System.out;
            this.error = System.err;

            // Prevent accidental writes since Bazel treats system streams as worker communication channel.
            System.setOut(new PrintStream(new DevNullOutputStream()));
            System.setErr(new PrintStream(new DevNullOutputStream()));
        }

        @Override
        public InputStream input() {
            return input;
        }

        @Override
        public PrintStream output() {
            return output;
        }

        @Override
        public PrintStream error() {
            return error;
        }

        private static class DevNullOutputStream extends OutputStream {
            @Override
            public void write(int writeByte) {
            }
        }
    }

}

File: detekt/wrapper/src/main/kotlin/io/buildfoundation/bazel/detekt/stream/Streams.kt
package io.buildfoundation.bazel.detekt.stream

import java.io.InputStream
import java.io.OutputStream
import java.io.PrintStream

data class Streams(val input: InputStream, val output: PrintStream, val error: PrintStream) {

    companion object {
        fun system(): Streams {
            val systemInput = System.`in`
            val systemOutput = System.out
            val systemError = System.err

            // Suppressing system singleton stream to avoid accidental output from the tool.
            System.setOut(PrintStream(DevNullOutputStream()))

            return Streams(systemInput, systemOutput, systemError)
        }
    }

    private class DevNullOutputStream : OutputStream() {
        override fun write(byte: Int) = Unit
    }
}